lib/command/run.rb

```diff
@@ -177,41 +172,54 @@ def create_runner_workload # rubocop:disable Metrics/MethodLength
       end
     end
 
-    def update_runner_workload # rubocop:disable Metrics/MethodLength
-      step("Updating runner workload '#{runner_workload}' based on '#{original_workload}'") do
+    def update_runner_workload # rubocop:disable Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+      step("Updating runner workload '#{runner_workload}' based on '#{original_workload}'") do # rubocop:disable Metrics/BlockLength
+        @expected_deployed_version = cp.cron_workload_deployed_version(runner_workload)
+        should_update = false
+
         _, original_container_spec = base_workload_specs(original_workload)
         spec, container_spec = base_workload_specs(runner_workload)
 
         # Override image if specified
         image = config.options[:image]
-        if image
-          image = latest_image if image == "latest"
-          @image_link = "/org/#{config.org}/image/#{image}"
-        else
-          @image_link = original_container_spec["image"]
+        image_link = if image
+                       image = latest_image if image == "latest"
+                       "/org/#{config.org}/image/#{image}"
+                     else
+                       original_container_spec["image"]
+                     end
+        if container_spec["image"] != image_link
+          container_spec["image"] = image_link
+          should_update = true
         end
-        @image_changed = container_spec["image"] != image_link
-        container_spec["image"] = image_link
 
         # Container overrides
-        container_spec["cpu"] = config.options[:cpu] if config.options[:cpu]
-        container_spec["memory"] = config.options[:memory] if config.options[:memory]
+        if config.options[:cpu] && container_spec["cpu"] != config.options[:cpu]
+          container_spec["cpu"] = config.options[:cpu]
+          should_update = true
+        end
+        if config.options[:memory] && container_spec["memory"] != config.options[:memory]
+          container_spec["memory"] = config.options[:memory]
+          should_update = true
+        end
+
+        next true unless should_update
 
         # Update runner workload
+        @expected_deployed_version += 1
         cp.apply_hash("kind" => "workload", "name" => runner_workload, "spec" => spec)
       end
     end
 
     def wait_for_runner_workload_create
       step("Waiting for runner workload '#{runner_workload}' to be created", retry_on_failure: true) do
-        cp.fetch_workload(runner_workload)
+        !cp.cron_workload_deployed_version(runner_workload).nil?
       end
     end
 
     def wait_for_runner_workload_update
       step("Waiting for runner workload '#{runner_workload}' to be updated", retry_on_failure: true) do
-        _, container_spec = base_workload_specs(runner_workload)
-        container_spec["image"] == image_link
+        cp.cron_workload_deployed_version(runner_workload) == expected_deployed_version
```

Comment:
This could be an endless loop if the workload is updated by someone else at the same time or if another `run` job is starting at the same time, because then the version would increase. Perhaps we should use `>=` instead of `==`. Then again, that wouldn't fix the issue of simultaneous jobs.

Reply:
Yes, 100%.

And log the situation. 

Maybe we're better off with the temp workload strategy b/c of the race condition?

Reply:
```suggestion
        cp.cron_workload_deployed_version(runner_workload) >= expected_deployed_version
```